File: vscode-extension/src/ide/settings/environmentDetection.ts
/**
 * Environment Detection
 * 
 * Utility functions for detecting Node.js and Codex installation paths and versions.
 * Translated from jetbrains-plugin/src/main/kotlin/com/asakii/settings/EnvironmentDetection.kt
 */

import * as os from 'os';
import * as path from 'path';
import * as fs from 'fs';
import { execSync } from 'child_process';

/**
 * Node.js detection result
 */
export interface NodeInfo {
    path: string;
    version?: string;
}

/**
 * Codex detection result
 */
export interface CodexInfo {
    path: string;
    version?: string;
}

/**
 * Environment detection utilities
 */
export const EnvironmentDetection = {
    /**
     * Detect Node.js path and version
     * @returns NodeInfo containing path and version, or undefined if not found
     */
    detectNodeInfo(): NodeInfo | undefined {
        const nodePath = this.detectNodePath();
        if (!nodePath) return undefined;

        const version = this.detectNodeVersion(nodePath);
        return { path: nodePath, version: version || undefined };
    },

    /**
     * Detect Codex path and version
     * @returns CodexInfo containing path and version, or undefined if not found
     */
    detectCodexInfo(): CodexInfo | undefined {
        const codexPath = this.detectCodexPath();
        if (!codexPath) return undefined;

        const version = this.detectCodexVersion(codexPath);
        return { path: codexPath, version: version || undefined };
    },

    /**
     * Auto-detect system Node.js path
     * Uses login shell execution to correctly load user's environment variables (PATH, etc.)
     * @returns Node.js executable path, empty string if not found
     */
    detectNodePath(): string {
        const isWindows = os.platform() === 'win32';

        // 1. Try to find via shell command (consistent with runtime logic)
        try {
            const command = isWindows ? 'where node' : 'which node';
            const result = execSync(command, { 
                encoding: 'utf8',
                stdio: ['pipe', 'pipe', 'pipe']
            }).trim().split('\n')[0];

            if (result && fs.existsSync(result)) {
                return result;
            }
        } catch {
            // Ignore error, continue trying other methods
        }

        // 2. Check common installation paths
        const commonPaths = isWindows
            ? [
                'C:\\Program Files\\nodejs\\node.exe',
                'C:\\Program Files (x86)\\nodejs\\node.exe',
                process.env.LOCALAPPDATA ? path.join(process.env.LOCALAPPDATA, 'Programs', 'node', 'node.exe') : null,
                process.env.APPDATA ? path.join(process.env.APPDATA, 'nvm', 'current', 'node.exe') : null,
                process.env.NVM_HOME ? path.join(process.env.NVM_HOME, 'current', 'node.exe') : null
            ]
            : [
                '/usr/local/bin/node',
                '/usr/bin/node',
                '/opt/homebrew/bin/node',
                process.env.HOME ? path.join(process.env.HOME, '.nvm', 'current', 'bin', 'node') : null,
                process.env.HOME ? path.join(process.env.HOME, '.local', 'bin', 'node') : null
            ];

        for (const nodePath of commonPaths) {
            if (nodePath && fs.existsSync(nodePath)) {
                return nodePath;
            }
        }

        return '';
    },

    /**
     * Auto-detect system Codex path
     * @returns Codex executable path, empty string if not found
     */
    detectCodexPath(): string {
        const isWindows = os.platform() === 'win32';
        const isMac = os.platform() === 'darwin';

        // Search PATH directly
        const pathEnv = process.env.PATH || '';
        const pathEntries = pathEnv.split(path.delimiter);
        const fileNames = isWindows
            ? ['codex.exe', 'codex.cmd', 'codex.bat']
            : ['codex'];

        for (const dir of pathEntries) {
            if (!dir) continue;
            for (const name of fileNames) {
                const candidate = path.join(dir, name);
                try {
                    if (fs.existsSync(candidate) && fs.statSync(candidate).isFile()) {
                        return candidate;
                    }
                } catch {
                    // Ignore access errors
                }
            }
        }

        // Common install paths
        const commonPaths: (string | null)[] = isWindows
            ? [
                'C:\\Program Files\\Codex\\codex.exe',
                'C:\\Program Files (x86)\\Codex\\codex.exe',
                process.env.LOCALAPPDATA ? path.join(process.env.LOCALAPPDATA, 'Codex', 'codex.exe') : null,
                process.env.USERPROFILE ? path.join(process.env.USERPROFILE, '.codex', 'codex.exe') : null
            ]
            : isMac
            ? [
                '/usr/local/bin/codex',
                '/opt/homebrew/bin/codex',
                os.homedir() ? path.join(os.homedir(), '.codex', 'codex') : null,
                '/Applications/Codex.app/Contents/MacOS/codex'
            ]
            : [
                '/usr/local/bin/codex',
                '/usr/bin/codex',
                os.homedir() ? path.join(os.homedir(), '.local', 'bin', 'codex') : null,
                os.homedir() ? path.join(os.homedir(), '.codex', 'codex') : null
            ];

        for (const codexPath of commonPaths) {
            if (codexPath) {
                try {
                    if (fs.existsSync(codexPath) && fs.statSync(codexPath).isFile()) {
                        return codexPath;
                    }
                } catch {
                    // Ignore access errors
                }
            }
        }

        // Vendor fallback (dev environment)
        const vendorPath = this.detectCodexVendorBinary();
        if (vendorPath) {
            return vendorPath;
        }

        return '';
    },

    /**
     * Detect Node.js version
     * @param nodePath Node.js executable path
     * @returns Version string (e.g., v24.2.0), null if not detected
     */
    detectNodeVersion(nodePath: string): string | null {
        try {
            const result = execSync(`"${nodePath}" --version`, {
                encoding: 'utf8',
                stdio: ['pipe', 'pipe', 'pipe']
            }).trim();

            if (result) {
                return result;
            }
        } catch {
            // Ignore error
        }

        return null;
    },

    /**
     * Detect Codex version
     * @param codexPath Codex executable path
     * @returns Version string, null if not detected
     */
    detectCodexVersion(codexPath: string): string | null {
        try {
            const result = execSync(`"${codexPath}" --version`, {
                encoding: 'utf8',
                stdio: ['pipe', 'pipe', 'pipe']
            }).trim();

            if (result) {
                return result;
            }
        } catch {
            // Ignore error
        }

        return null;
    },

    /**
     * Detect Codex vendor binary in development environment
     * @returns Vendor path, null if not found
     */
    detectCodexVendorBinary(): string | null {
        const platform = os.platform();
        const arch = os.arch();
        const isWindows = platform === 'win32';

        let triple: string | null = null;
        if (isWindows && arch.includes('64')) {
            triple = 'x86_64-pc-windows-msvc';
        } else if (platform === 'darwin' && arch === 'arm64') {
            triple = 'aarch64-apple-darwin';
        } else if (platform === 'darwin' && arch.includes('x86')) {
            triple = 'x86_64-apple-darwin';
        } else if (platform === 'linux' && arch === 'arm64') {
            triple = 'aarch64-unknown-linux-musl';
        } else if (platform === 'linux' && arch.includes('64')) {
            triple = 'x86_64-unknown-linux-musl';
        }

        if (!triple) return null;

        const binaryName = isWindows ? 'codex.exe' : 'codex';
        const candidate = path.join(
            'external',
            'openai-codex',
            'sdk',
            'vendor',
            triple,
            'codex',
            binaryName
        );

        try {
            if (fs.existsSync(candidate) && fs.statSync(candidate).isFile()) {
                return path.resolve(candidate);
            }
        } catch {
            // Ignore access errors
        }

        return null;
    },

    /**
     * Check if current platform is Windows
     */
    isWindows(): boolean {
        return os.platform() === 'win32';
    },

    /**
     * Check if current platform is macOS
     */
    isMac(): boolean {
        return os.platform() === 'darwin';
    },

    /**
     * Check if current platform is Linux
     */
    isLinux(): boolean {
        return os.platform() === 'linux';
    }
};

export default EnvironmentDetection;

File: jetbrains-plugin/src/main/kotlin/com/asakii/settings/EnvironmentDetection.kt
package com.asakii.settings

import java.io.File
import java.nio.file.Paths

/**
 * Claude CLI 检测结果
 */
data class ClaudeCliInfo(
    val path: String,
    val version: String? = null
)

/**
 * Node.js 检测结果
 */
data class NodeInfo(
    val path: String,
    val version: String? = null
)

/**
 * Codex 检测结果
 */
data class CodexInfo(
    val path: String,
    val version: String? = null
)

/**
 * 环境检测工具类
 *
 * 用于检测系统中的 Claude CLI、Node.js 和 Codex 安装路径及版本
 */
object EnvironmentDetection {

    /**
     * 检测 Claude CLI 路径和版本
     * @return ClaudeCliInfo 包含路径和版本，未找到返回 null
     */
    fun detectClaudeInfo(): ClaudeCliInfo? {
        val path = detectClaudePath()
        if (path.isEmpty()) return null

        val version = detectClaudeVersion(path)
        return ClaudeCliInfo(path, version)
    }

    /**
     * 自动检测系统中的 Claude CLI 路径
     * @return Claude CLI 可执行文件路径，未找到返回空字符串
     */
    fun detectClaudePath(): String {
        val isWindows = System.getProperty("os.name").lowercase().contains("win")

        // 1. 通过 login shell 查找
        try {
            val command = if (isWindows) {
                // Windows: 查找 claude.cmd（npm 安装的可执行包装脚本）
                arrayOf("cmd", "/c", "where", "claude.cmd")
            } else {
                val defaultShell = System.getenv("SHELL") ?: "/bin/bash"
                arrayOf(defaultShell, "-l", "-c", "which claude")
            }

            val process = ProcessBuilder(*command)
                .redirectErrorStream(true)
                .start()

            val result = process.inputStream.bufferedReader().readLine()?.trim()
            val exitCode = process.waitFor()

            if (exitCode == 0 && !result.isNullOrBlank()) {
                return result
            }
        } catch (_: Exception) {
            // 忽略错误，继续尝试常见路径
        }

        // 2. 检查常见安装路径
        val commonPaths = if (isWindows) {
            listOf(
                System.getenv("APPDATA")?.let { "$it\\npm\\claude.cmd" },
                System.getenv("LOCALAPPDATA")?.let { "$it\\npm\\claude.cmd" },
                System.getenv("USERPROFILE")?.let { "$it\\.npm-global\\claude.cmd" },
                "C:\\Program Files\\nodejs\\claude.cmd"
            )
        } else {
            listOf(
                "/usr/local/bin/claude",
                "/usr/bin/claude",
                "/opt/homebrew/bin/claude",
                System.getenv("HOME")?.let { "$it/.npm-global/bin/claude" },
                System.getenv("HOME")?.let { "$it/.local/bin/claude" }
            )
        }

        for (path in commonPaths) {
            if (path != null && File(path).exists()) {
                return path
            }
        }

        return ""
    }

    /**
     * 检测 Claude CLI 版本
     */
    private fun detectClaudeVersion(claudePath: String): String? {
        try {
            // Windows 上 .cmd 文件需要通过 cmd /c 启动；Unix 直接执行
            val command = if (claudePath.lowercase().endsWith(".cmd")) {
                arrayOf("cmd", "/c", claudePath, "--version")
            } else {
                arrayOf(claudePath, "--version")
            }

            val process = ProcessBuilder(*command)
                .redirectErrorStream(true)
                .start()

            val result = process.inputStream.bufferedReader().readLine()?.trim()
            val exitCode = process.waitFor()

            if (exitCode == 0 && !result.isNullOrBlank()) {
                return result
            }
        } catch (_: Exception) {
            // 忽略错误
        }

        return null
    }

    /**
     * 检测 Node.js 路径和版本
     * @return NodeInfo 包含路径和版本，未找到返回 null
     */
    fun detectNodeInfo(): NodeInfo? {
        val path = detectNodePath()
        if (path.isEmpty()) return null

        val version = detectNodeVersion(path)
        return NodeInfo(path, version)
    }

    /**
     * 检测 Codex 路径和版本
     * @return CodexInfo 包含路径和版本，未找到返回 null
     */
    fun detectCodexInfo(): CodexInfo? {
        val path = detectCodexPath()
        if (path.isEmpty()) return null

        val version = detectCodexVersion(path)
        return CodexInfo(path, version)
    }

    /**
     * 自动检测系统中的 Node.js 路径
     * 使用 login shell 执行，以正确加载用户的环境变量（PATH 等）
     * @return Node.js 可执行文件路径，未找到返回空字符串
     */
    fun detectNodePath(): String {
        val isWindows = System.getProperty("os.name").lowercase().contains("win")

        // 1. 尝试通过 login shell 查找（与运行时逻辑一致）
        try {
            val command = if (isWindows) {
                // Windows: 使用 cmd /c
                arrayOf("cmd", "/c", "where", "node")
            } else {
                // macOS/Linux: 使用 login shell 执行 which node
                val defaultShell = System.getenv("SHELL") ?: "/bin/bash"
                arrayOf(defaultShell, "-l", "-c", "which node")
            }

            val process = ProcessBuilder(*command)
                .redirectErrorStream(true)
                .start()

            val result = process.inputStream.bufferedReader().readLine()?.trim()
            val exitCode = process.waitFor()

            if (exitCode == 0 && !result.isNullOrBlank() && File(result).exists()) {
                return result
            }
        } catch (_: Exception) {
            // 忽略错误，继续尝试其他方式
        }

        // 2. 检查常见安装路径
        val commonPaths = if (isWindows) {
            listOf(
                "C:\\Program Files\\nodejs\\node.exe",
                "C:\\Program Files (x86)\\nodejs\\node.exe",
                System.getenv("LOCALAPPDATA")?.let { "$it\\Programs\\node\\node.exe" },
                System.getenv("APPDATA")?.let { "$it\\nvm\\current\\node.exe" },
                System.getenv("NVM_HOME")?.let { "$it\\current\\node.exe" }
            )
        } else {
            listOf(
                "/usr/local/bin/node",
                "/usr/bin/node",
                "/opt/homebrew/bin/node",
                System.getenv("HOME")?.let { "$it/.nvm/current/bin/node" },
                System.getenv("HOME")?.let { "$it/.local/bin/node" }
            )
        }

        for (path in commonPaths) {
            if (path != null && File(path).exists()) {
                return path
            }
        }

        return ""
    }

    /**
     * 自动检测系统中的 Codex 路径
     * @return Codex 可执行文件路径，未找到返回空字符串
     */
    fun detectCodexPath(): String {
        val os = System.getProperty("os.name").lowercase()
        val isWindows = os.contains("win")

        // Search PATH directly
        val pathEnv = System.getenv("PATH") ?: ""
        val pathEntries = pathEnv.split(File.pathSeparatorChar)
        val fileNames = if (isWindows) {
            listOf("codex.exe", "codex.cmd", "codex.bat")
        } else {
            listOf("codex")
        }
        for (dir in pathEntries) {
            if (dir.isBlank()) continue
            for (name in fileNames) {
                val candidate = File(dir, name)
                if (candidate.exists() && candidate.canExecute()) {
                    return candidate.absolutePath
                }
            }
        }

        // Common install paths
        val commonPaths = when {
            isWindows -> listOf(
                "C:\\Program Files\\Codex\\codex.exe",
                "C:\\Program Files (x86)\\Codex\\codex.exe",
                System.getenv("LOCALAPPDATA")?.let { "$it\\Codex\\codex.exe" },
                System.getenv("USERPROFILE")?.let { "$it\\.codex\\codex.exe" }
            )
            os.contains("mac") -> listOf(
                "/usr/local/bin/codex",
                "/opt/homebrew/bin/codex",
                System.getProperty("user.home")?.let { "$it/.codex/codex" },
                "/Applications/Codex.app/Contents/MacOS/codex"
            )
            else -> listOf(
                "/usr/local/bin/codex",
                "/usr/bin/codex",
                System.getProperty("user.home")?.let { "$it/.local/bin/codex" },
                System.getProperty("user.home")?.let { "$it/.codex/codex" }
            )
        }

        for (path in commonPaths) {
            if (path != null) {
                val file = File(path)
                if (file.exists() && file.canExecute()) {
                    return file.absolutePath
                }
            }
        }

        // Vendor fallback (dev environment)
        val vendorPath = detectCodexVendorBinary()
        if (vendorPath != null) {
            return vendorPath
        }

        return ""
    }

    /**
     * 检测 Node.js 版本
     * @param nodePath Node.js 可执行文件路径
     * @return 版本号（如 v24.2.0），未检测到返回 null
     */
    private fun detectNodeVersion(nodePath: String): String? {
        val isWindows = System.getProperty("os.name").lowercase().contains("win")

        try {
            val command = if (isWindows) {
                arrayOf("cmd", "/c", nodePath, "--version")
            } else {
                val defaultShell = System.getenv("SHELL") ?: "/bin/bash"
                arrayOf(defaultShell, "-l", "-c", "$nodePath --version")
            }

            val process = ProcessBuilder(*command)
                .redirectErrorStream(true)
                .start()

            val result = process.inputStream.bufferedReader().readLine()?.trim()
            val exitCode = process.waitFor()

            if (exitCode == 0 && !result.isNullOrBlank()) {
                return result
            }
        } catch (_: Exception) {
            // 忽略错误
        }

        return null
    }

    /**
     * 检测 Codex 版本
     * @param codexPath Codex 可执行文件路径
     * @return 版本号，未检测到返回 null
     */
    private fun detectCodexVersion(codexPath: String): String? {
        val isWindows = System.getProperty("os.name").lowercase().contains("win")

        try {
            val command = if (isWindows) {
                arrayOf("cmd", "/c", codexPath, "--version")
            } else {
                arrayOf(codexPath, "--version")
            }

            val process = ProcessBuilder(*command)
                .redirectErrorStream(true)
                .start()

            val result = process.inputStream.bufferedReader().readLine()?.trim()
            val exitCode = process.waitFor()

            if (exitCode == 0 && !result.isNullOrBlank()) {
                return result
            }
        } catch (_: Exception) {
            // ignore
        }

        return null
    }

    /**
     * 检测开发环境中的 Codex vendor 二进制文件
     * @return vendor 路径，未找到返回 null
     */
    private fun detectCodexVendorBinary(): String? {
        val os = System.getProperty("os.name").lowercase()
        val arch = System.getProperty("os.arch").lowercase()
        val isWindows = os.contains("win")

        val triple = when {
            isWindows && arch.contains("64") -> "x86_64-pc-windows-msvc"
            os.contains("mac") && arch.contains("aarch64") -> "aarch64-apple-darwin"
            os.contains("mac") && arch.contains("x86") -> "x86_64-apple-darwin"
            os.contains("linux") && arch.contains("aarch64") -> "aarch64-unknown-linux-musl"
            os.contains("linux") && arch.contains("64") -> "x86_64-unknown-linux-musl"
            else -> null
        } ?: return null

        val binaryName = if (isWindows) "codex.exe" else "codex"
        val candidate = Paths.get(
            "external",
            "openai-codex",
            "sdk",
            "vendor",
            triple,
            "codex",
            binaryName
        ).toFile()

        return if (candidate.exists() && candidate.canExecute()) {
            candidate.absolutePath
        } else {
            null
        }
    }
}
